sklearn/model_selection/_split.py

```diff
@@ -572,17 +573,29 @@ def __init__(self, n_splits=3, shuffle=False, random_state=None):
 
     def _make_test_folds(self, X, y=None):
         rng = self.random_state
-        y = np.asarray(y)
-        type_of_target_y = type_of_target(y)
+
+        # For compatibility with GridSearchCV (which accepts y to be a sparse
+        # matrix), if y is sparse we convert it to a dense matrix. An error is
+        # raised if y does not have the right shape, to avoid creating a large
+        # dense matrix (whose shape will raise an error in ``column_or_1d``).
+        if sp.issparse(y):
+            shape = y.shape
+            if len(shape) > 2 or (len(shape) == 2 and shape[1] != 1):
+                raise ValueError("bad input shape {0}".format(shape))
+            else:
+                _y = y.todense()
```

Comment:
Use "toarray" rather than "todense", to avoid creating a matrix instead of an array.